pkg/helpers/helpers.go

```diff
@@ -0,0 +1,641 @@
+	}
+	return isV1 == -1
+}
+
+// MustCreateObjectFromTemplate render a template to a runtime object with its configuration
+// If it's failed, this function will panic
+func MustCreateObjectFromTemplate(file string, template []byte, config interface{}) runtime.Object {
+	raw := MustCreateAssetFromTemplate(file, template, config)
+	return mustCreateObject(raw)
+}
+
+// MustCreateAssetFromTemplate render a template with its configuration
+// If it's failed, this function will panic
+func MustCreateAssetFromTemplate(name string, tb []byte, config interface{}) []byte {
+	tmpl, err := template.New(name).Parse(string(tb))
+	if err != nil {
+		panic(err)
+	}
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, config); err != nil {
+		panic(err)
+	}
+	return buf.Bytes()
+}
+
+// ManifestsEqual if two manifests are equal, return true
+func ManifestsEqual(newManifests, oldManifests []workv1.Manifest) bool {
+	if len(newManifests) != len(oldManifests) {
+		return false
+	}
+
+	for i := range newManifests {
+		if !equality.Semantic.DeepEqual(newManifests[i].Raw, oldManifests[i].Raw) {
+			return false
+		}
+	}
+	return true
+}
+
+// ApplyResources apply resources, includes: serviceaccount, secret, deployment, clusterrole, clusterrolebinding,
+// crdv1beta1, crdv1, manifestwork and klusterlet
+func ApplyResources(client client.Client, recorder record.EventRecorder,
+	scheme *runtime.Scheme, owner metav1.Object, objs ...runtime.Object) error {
+	errs := []error{}
+	for _, obj := range objs {
+		if owner != nil {
+			required, ok := obj.(metav1.Object)
+			if !ok {
+				errs = append(errs, fmt.Errorf("%T is not a metav1.Object, cannot call SetControllerReference", obj))
+				continue
+			}
+
+			if err := controllerutil.SetControllerReference(owner, required, scheme); err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
+
+		switch required := obj.(type) {
+		case *corev1.ServiceAccount:
+			errs = append(errs, applyServiceAccount(client, recorder, required))
+		case *corev1.Secret:
+			errs = append(errs, applySecret(client, recorder, required))
+		case *appsv1.Deployment:
+			errs = append(errs, applyDeployment(client, recorder, required))
+		case *rbacv1.ClusterRole:
+			errs = append(errs, applyClusterRole(client, recorder, required))
+		case *rbacv1.ClusterRoleBinding:
+			errs = append(errs, applyClusterRoleBinding(client, recorder, required))
+		case *crdv1beta1.CustomResourceDefinition:
+			errs = append(errs, applyCRDV1beta1(client, recorder, required))
+		case *crdv1.CustomResourceDefinition:
+			errs = append(errs, applyCRDV1(client, recorder, required))
+		case *workv1.ManifestWork:
+			errs = append(errs, applyManifestWork(client, recorder, required))
+		case *operatorv1.Klusterlet:
+			errs = append(errs, applyKlusterlet(client, recorder, required))
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func applyServiceAccount(client client.Client, recorder record.EventRecorder, required *corev1.ServiceAccount) error {
```

Comment:
why we need these apply funcs, couldn't we use library-go?

Reply:
using library-go, we need adapt the `library-go` argument types with controller-runtime, but I can try use library-go